odoo/tests/common.py

```diff
@@ -1960,17 +1991,77 @@ def setUp(self):
 
         self._logger = self._logger.getChild(self._testMethodName)
 
-        self.xmlrpc_common = xmlrpclib.ServerProxy(self.xmlrpc_url + 'common', transport=Transport(self.cr))
-        self.xmlrpc_db = xmlrpclib.ServerProxy(self.xmlrpc_url + 'db', transport=Transport(self.cr))
-        self.xmlrpc_object = xmlrpclib.ServerProxy(self.xmlrpc_url + 'object', transport=Transport(self.cr), use_datetime=True)
+        _self = self
+        class HttpCaseXmlrpcProxy(xmlrpclib.ServerProxy):
+            # ServerProxy's internal uses some weird magic..
+            def _ServerProxy__request(self, *args, **kwargs):
+                with _self.allow_requests(all_requests=True):
+                    return super()._ServerProxy__request(*args, **kwargs)
+
+        self.xmlrpc_common = HttpCaseXmlrpcProxy(self.xmlrpc_url + 'common', transport=Transport(self.cr))
+        self.xmlrpc_db = HttpCaseXmlrpcProxy(self.xmlrpc_url + 'db', transport=Transport(self.cr))
+        self.xmlrpc_object = HttpCaseXmlrpcProxy(self.xmlrpc_url + 'object', transport=Transport(self.cr), use_datetime=True)
         # setup an url opener helper
         self.opener = Opener(self.cr)
+        # we need to allow requests during pdf rendering.
+        old_run_wkhtmltopdf = ir_actions_report._run_wkhtmltopdf
+        def _patched_run_wkhtmltopdf(*args):
+            with self.allow_requests(all_requests=True):
+                return old_run_wkhtmltopdf(*args)
+        self.startPatcher(
+            patch.object(ir_actions_report, '_run_wkhtmltopdf', _patched_run_wkhtmltopdf)
+        )
 
     @contextmanager
     def enter_registry_test_mode(self):
         _logger.warning("HTTPCase is already in test mode")
         yield
 
+    @contextmanager
+    def allow_pdf_render(self):
+        _logger.warning("HTTPCase does not require calling allow_pdf_render")
+        yield
+
+    @contextmanager
+    def allow_requests(self, browser: ChromeBrowser | None = None, all_requests=False):
+        """
+        Allows HTTP requests for the scope of the context.
+
+        Params:
+            browser (ChromeBrowser | None): if given, add the cookie to the browser.
+            all_requests (bool): if True, allows all requests regardless of cookie.
+        """
+        assert hasattr(self, '_registry_test_lock')
+        old_lock_id = test_cursor.request_lock_id
+        previous_cookie = self.opener.cookies.get(test_cursor.TEST_CURSOR_COOKIE_NAME)
+        released = False
+        try:
+            try:
+                self._registry_test_lock.release()
+                released = True
+            except RuntimeError:
+                # It is possible to have two nested calls, in which case we
+                # might not have the lock acquired but we can simply ignore it.
+                pass
+            lock_id = f'{self._testMethodName}__{str(next(_http_cursor_lock_id_seq))}'
```

Comment:
Actually having a global lock count is not that useful, it's easier to track where we are for failing requests if the sequence is localized to the test run itself.